src/NUnitEngine/nunit.engine.core/Drivers/NUnitFrameworkApi2018.cs

```diff
@@ -0,0 +1,253 @@
+﻿// Copyright (c) Charlie Poole, Rob Prouse and Contributors. MIT License - see LICENSE.txt
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Reflection;
+using System.Runtime.Serialization;
+using NUnit.Common;
+using NUnit.Engine.Internal;
+
+namespace NUnit.Engine.Drivers
+{
+    /// <summary>
+    /// This is the revised API, designed for use with .NET Core. It first
+    /// appears in our source code in 2018. This implementation is modified
+    /// to make it work under the .NET Framework as well as .NET Core. It
+    /// may be used for NUnit 3.10 or higher.
+    /// </summary>
+#if NETFRAMEWORK
+    public class NUnitFrameworkApi2018 : MarshalByRefObject, NUnitFrameworkApi
+#else
+    public class NUnitFrameworkApi2018 : NUnitFrameworkApi
+#endif
+    {
+        static readonly ILogger log = InternalTrace.GetLogger(nameof(NUnitFrameworkApi2018));
+
+        const string LOAD_MESSAGE = "Method called without calling Load first. Possible error in runner.";
+        const string INVALID_FRAMEWORK_MESSAGE = "Running tests against this version of the framework using this driver is not supported. Please update NUnit.Framework to the latest version.";
+        const string FAILED_TO_LOAD_ASSEMBLY = "Failed to load assembly ";
+        const string FAILED_TO_LOAD_NUNIT = "Failed to load the NUnit Framework in the test assembly";
+
+        const string CONTROLLER_TYPE = "NUnit.Framework.Api.FrameworkController";
+
+        private readonly string _driverId;
+
+        private readonly AssemblyName _nunitRef;
+
+        private object? _frameworkController;
+        private Type? _frameworkControllerType;
+
+#if NETCOREAPP
+        private TestAssemblyLoadContext? _assemblyLoadContext;
+        //private Assembly? _testAssembly;
+        private Assembly? _frameworkAssembly;
+#endif
+
+        private string? _testAssemblyPath;
+
+        public NUnitFrameworkApi2018(string driverId, AssemblyName nunitRef)
+        {
+            Guard.ArgumentNotNull(driverId, nameof(driverId));
+            Guard.ArgumentNotNull(nunitRef, nameof(nunitRef));
+
+            _driverId = driverId;
+            _nunitRef = nunitRef;
+        }
+
+        public string Load(string testAssemblyPath, IDictionary<string, object> settings)
+        {
+            Guard.ArgumentNotNull(testAssemblyPath, nameof(testAssemblyPath));
+            Guard.ArgumentNotNull(settings, nameof(settings));
+            Guard.ArgumentValid(File.Exists(testAssemblyPath), "Framework driver called with a file name that doesn't exist.", "testAssemblyPath");
+            log.Info($"Loading {testAssemblyPath} - see separate log file");
+
+            _testAssemblyPath = Path.GetFullPath(testAssemblyPath);
+            var idPrefix = string.IsNullOrEmpty(_driverId) ? "" : _driverId + "-";
```

Comment:
`_driverId` should never be `null` and with a change to `Guard.IsNotNullOrEmpty` in the constructor it should not be empty either. We can then remove this conditional here.